### ISSUE TITLE
Asynchronous HTTP progress report & Pause/Resume downloads

### ISSUE THREAD

**Issue by CaptainPRICE** [Facepunch/garrysmod-requests#1340]
I would like to suggest to add a 'progress callback' to [HTTP](https://wiki.garrysmod.com/page/Global/HTTP).
1. Expose a couple of variables via 'progress callback', such as **total file size** (in bytes), **transferred size** (downloaded/uploaded size *so far*, in bytes), **transfer rate** (download/upload speed, in bytes/sec). And perhaps also approximate time left (in seconds), so that we don't have to do calculation manually.
2. There's no way of stopping ~~or even pausing/resuming~~ a download/upload. Please consider implementing it.

I just tried to make HTTP downloader/uploader addon, but right now there's nothing more than 'completion callback', alone, is pretty limited functionality. I was aiming for something such as:
![image](https://user-images.githubusercontent.com/9789070/57919679-94020300-7899-11e9-92ea-dc0535a318c0.png)

P.S. I know you may get the total file size by sending a [HTTP HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) request, but the thing is, a file might change between the two separate HTTP requests.

**Comment by CornerPin** [comment]
It is not possible to get the file size if server returns [chunked data](https://en.wikipedia.org/wiki/Chunked_transfer_encoding). Also, _I think_ if you pause too much, the server will close the connection because of the time out.

**Comment by CaptainPRICE** [comment]
> Also, _I think_ if you pause too much, the server will close the connection because of the time out.

Yes. That's why there is "Resume capability" indication (see screenshot).

---

This makes me think, it would be **best** to have a class that would represent a HTTP file download/upload request, and a separate class to represent a response.

Please consider: [**API Draft**](https://gist.github.com/CaptainPRICE/e1cea02f2ab2a4ae24b700f9b238f4b8).

I have spent some time to write that document. Thoughts? :)
P.S. Bear with me. English is not my native language.

**Comment by robotboy655** [comment]
I can see how some sort of ability to get download progress can be useful, but pausing simply cannot be added, we are using Steamworks HTTP ( https://partner.steamgames.com/doc/api/ISteamHTTP ) stuff and it doesn't have pausing.

**Comment by CaptainPRICE** [comment]
@robotboy655 I have updated API draft, please [review the changes](https://gist.github.com/CaptainPRICE/e1cea02f2ab2a4ae24b700f9b238f4b8#gistcomment-2921359).
Is it okay? Is there anything else to drop or possibly add?
